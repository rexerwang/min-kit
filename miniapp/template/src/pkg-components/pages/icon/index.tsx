import { MinIcon } from '@min-kit/components'
import { copy } from '@min-kit/extends'
import { View } from '@tarojs/components'

import MinCode from '@/components/code'
import Layout from '@/components/layout'

// pages/index/index.tsx
export default function Index() {
  return (
    <>
      <Layout title='MinIcon'>
        <View className='grid grid-cols-4 gap-4'>
          {Object.keys(MinIcon.configs).map((name) => (
            <View key={name} className='flex-center flex-col'>
              <View className='p-1 rounded' hoverClass='bg-gray-900' onClick={() => copy(`<MinIcon name='${name}' />`)}>
                <MinIcon name={name} size={32} />
              </View>
              <View className='text-xs'>{name}</View>
            </View>
          ))}
        </View>
      </Layout>

      <MinCode.Layout className='mt-4' lang='tsx'>{` // load icons at first
MinIcon.load({
  'react': require('./react.svg')
})

<MinIcon name='react' />`}</MinCode.Layout>
    </>
  )
}

definePageConfig({
  navigationBarTitleText: 'MinIcon',
})
