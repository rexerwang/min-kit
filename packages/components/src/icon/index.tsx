import { isString, px } from '@min-kit/shared'
import { Image, ImageProps, View, ViewProps } from '@tarojs/components'
import clsx from 'clsx'
import { useMemo } from 'react'

export interface MinIconProps extends Omit<ImageProps, 'src' | 'mode'> {
  name: string
  /** `size` or `[width, height]` */
  size?: number | number[]
  style?: React.CSSProperties
}

export function MinIcon({ name, size = 36, ...props }: MinIconProps) {
  const style = useMemo(() => {
    let width: number
    let height: number

    if (Array.isArray(size)) {
      width = size[0]
      height = size[1]
    } else {
      width = size
      height = size
    }

    return { ...px.of({ width, height }), ...props.style }
  }, [props.style, size])

  return <Image {...props} src={MinIcon.configs[name]} style={style} />
}

MinIcon.displayName = 'MinIcon'

interface MinIconFontProps extends ViewProps {
  name: string
}

function IconFont({ name, className, ...props }: MinIconFontProps) {
  return (
    <View
      {...props}
      className={clsx('min-icon-font iconfont icon-' + name, className)}
      hoverClass={props.onClick ? props.hoverClass ?? 'min-icon-font--hover' : undefined}
    />
  )
}

IconFont.displayName = 'MinIconFont'
MinIcon.Font = IconFont

MinIcon.configs = {} as Record<string, string>
MinIcon.load = (icons: Record<string, string>) => {
  for (const name in icons) {
    const path = icons[name]
    if (path && isString(path)) {
      MinIcon.configs[name] = path
    }
  }
}
