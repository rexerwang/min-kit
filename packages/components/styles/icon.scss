.min-icon-font {
  &.iconfont {
    font-family: iconfont !important;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  &--hover {
    @apply brightness-90;
  }
}
