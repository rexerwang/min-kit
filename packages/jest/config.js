const defineJestConfig = require('@tarojs/test-utils-react/dist/jest.js').default

module.exports = defineJestConfig({
  passWithNoTests: true,
  testEnvironment: 'jsdom',
  testMatch: ['<rootDir>/src/__tests__/**/*.test.ts?(x)'],
  collectCoverageFrom: ['<rootDir>/src/**/*.ts?(x)', '!**/__tests__/**', '!**/{debounce,throttle}.ts'],
  coverageReporters: ['clover', 'json', 'lcov', 'text-summary'],
  globals: {
    wx: {},
    tt: {},
    my: {},
  },
})
