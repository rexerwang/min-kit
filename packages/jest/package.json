{
  "name": "@min-kit/jest",
  "version": "0.5.11",
  "private": true,
  "description": "jest config & utils for internal",
  "author": "rexerwang",
  "license": "MIT",
  "repository": "rexerwang/min-kit",
  "exports": {
    ".": "./utils.ts",
    "./config": "./config.js"
  },
  "devDependencies": {
    "@happy-dom/jest-environment": "^14.7.1",
    "@tarojs/test-utils-react": "^0.1.1",
    "@testing-library/jest-dom": "^6.4.8",
    "@testing-library/react": "^16.0.1"
  }
}
