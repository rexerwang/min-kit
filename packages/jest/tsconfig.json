{
  "extends": "../../tsconfig.root.json",
  "include": ["utils.ts"]
}
