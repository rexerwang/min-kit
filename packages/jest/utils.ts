import TestUtils from '@tarojs/test-utils-react'

export const createTaroTestUtils = () => {
  const testUtils = new TestUtils()
  return Object.assign(testUtils, { unmount: testUtils.unmout })
}

export const spyOnConsole = (methods = ['error', 'warn', 'info', 'debug']) =>
  methods.map((method: any) => {
    const spy = jest.spyOn(console, method)
    spy.mockImplementation(jest.fn())
    return spy
  })

export * from '@testing-library/react'
