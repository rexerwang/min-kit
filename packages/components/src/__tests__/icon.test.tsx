import { createTaroTestUtils, render } from '@min-kit/jest'

import { MinIcon } from '../icon'

describe('<MinIcon />', () => {
  const taro = createTaroTestUtils()

  beforeAll(() => {
    MinIcon.load({ test: '/test.jpg' })
  })

  it('should render <MinIcon /> toMatchSnapshot', () => {
    expect(render(<MinIcon className='test' name='test' />).asFragment()).toMatchSnapshot()
    expect(render(<MinIcon name='test' size={[20, 40]} />).asFragment()).toMatchSnapshot()
  })

  it('should render <MinIcon.Font /> toMatchSnapshot', () => {
    const { asFragment } = render(<MinIcon.Font className='test' name='test' />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should MinIcon.Font dispatch click event', async () => {
    const onClick = jest.fn()
    await taro.mount(MinIcon.Font, { props: { name: 'test', onClick } })

    await taro.act(() => {
      taro.fireEvent.click(taro.queries.querySelector('.min-icon-font')!)
    })
    expect(onClick).toHaveBeenCalled()
  })

  it('should load icons correctly', () => {
    MinIcon.load({ test: '/a.jpg', b: '/b.jpg' })
    expect(MinIcon.configs).toEqual({ test: '/a.jpg', b: '/b.jpg' })
  })
})
