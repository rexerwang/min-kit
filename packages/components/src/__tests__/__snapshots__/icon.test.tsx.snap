// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MinIcon /> should render <MinIcon /> toMatchSnapshot 1`] = `
<DocumentFragment>
  <taro-image-core
    class="test"
    src="/test.jpg"
    style="width: 16px; height: 16px;"
  />
</DocumentFragment>
`;

exports[`<MinIcon /> should render <MinIcon /> toMatchSnapshot 2`] = `
<DocumentFragment>
  <taro-image-core
    class=""
    src="/test.jpg"
    style="width: 20px; height: 40px;"
  />
</DocumentFragment>
`;

exports[`<MinIcon /> should render <MinIcon.Font /> toMatchSnapshot 1`] = `
<DocumentFragment>
  <taro-view-core
    class="min-icon-font iconfont icon-test test"
  >
    <!---->
  </taro-view-core>
</DocumentFragment>
`;
